widget/src/IrcApp.tsx

```diff
@@ -0,0 +1,369 @@
+
+    const onChannelChange: React.ChangeEventHandler<HTMLInputElement> = useCallback((e) => {
+        let _channel = e.currentTarget.value;
+        _channel = _channel.startsWith('#') ? _channel : `#${_channel}`;
+        setChannel(_channel);
+    }, []);
+
+    const onOperatorNickChange: React.ChangeEventHandler<HTMLInputElement> = useCallback((e) => {
+        setOperatorNick(e.currentTarget.value);
+    }, []);
+
+    const onChannelKeyChange: React.ChangeEventHandler<HTMLInputElement> = useCallback((e) => {
+        setChannelKey(e.currentTarget.value);
+    }, []);
+
+    return <div className="grid grid-cols-1 gap-4 my-2">
+        <Forms.Input
+            label="Channel"
+            comment="Entering a channel will cause the bot to join it"
+            placeholder="#"
+            type="text"
+            value={channel}
+            onChange={onChannelChange}
+            onBlur={getOperatorNicks}
+            disabled={isBusy}
+        />
+        <Forms.Input
+            label="Channel operator nick"
+            comment={isBusyOperatorNicks ? 'Loading operator nicks...' : operatorNicksInfo}
+            type="text"
+            value={operatorNick}
+            list="operatorNicks"
+            onChange={onOperatorNickChange}
+            disabled={isBusy}
+        />
+        { operatorNicks &&
+            <datalist id="operatorNicks">
+                { operatorNicks.map(nick => <option value={nick} key={nick}></option>) }
+            </datalist>
+        }
+        <Forms.Input
+            label="Channel key (optional)"
+            type="text"
+            value={channelKey}
+            onChange={onChannelKeyChange}
+            onBlur={getOperatorNicks}
+            disabled={isBusy}
+        />
+        <Buttons.Primary onClick={linkChannel} disabled={!isFormValid || isBusy}>
+            Request link
+        </Buttons.Primary>
+        { error && <Alerts.Danger>{ error }</Alerts.Danger> }
+        { info && <Alerts.Success>{ info }</Alerts.Success> }
+    </div>;
+}
+
+const AvailableChannels = ({
+    client,
+    roomId,
+}: {
+    client: IrcProvisioningClient
+    roomId: string,
+}) => {
+    const [error, setError] = useState('');
+
+    const [networks, setNetworks] = useState<QueryNetworksResponse>();
+
+    useEffect(() => {
+        const getNetworks = async() => {
+            try {
+                const _networks = await client.queryNetworks();
+                setNetworks(_networks);
+            }
+            catch (e) {
+                console.error(e);
+                setError(
+                    'Could not get networks.'
+                    + ` ${e instanceof ProvisioningError ? e.message : ''}`
+                );
+            }
+        };
+        getNetworks();
+    }, [client]);
+
+    const [selectedServer, setSelectedServer] = useState('');
+
+    let content;
+    if (networks) {
+        if (networks.servers.length > 0) {
```

Comment:
My cheeky suggestion is that if the server length is 1 (which is true in nearly all cases), we should just render this as a readonly field. Something to make it look like not a dropdown.